### ISSUE TITLE
[Bug] NullReferenceException during Invoke in _updateModsList

### ISSUE THREAD

**Issue by Arkerthan** [KSP-CKAN/CKAN#3312]
Background
----------
* Operating System: Windows 10
* CKAN Version: 1.30.0
* KSP Version: 1.11.1.3066

**Have you made any manual changes to your GameData folder (i.e., not via CKAN)?**
No

Problem
-------
**Describe the bug**
During startup there is a  Microsoft .NET Framework exception. First startup after automatic CKAN update, since then every time.

Clicking `Continue` (`Weiter`) closes the exception window, but CKAN won't change even after several minutes of waiting.

**Steps to reproduce**
* Execute ckan.exe

**Expected behavior**
No exception

**Screenshots (if applicable)**
![image](https://user-images.githubusercontent.com/52896080/110208034-143a9380-7e87-11eb-9686-17c379f6f3f2.png)

**Full Stacktrace**
```
Informationen über das Aufrufen von JIT-Debuggen
anstelle dieses Dialogfelds finden Sie am Ende dieser Meldung.

************** Ausnahmetext **************
System.NullReferenceException: Der Objektverweis wurde nicht auf eine Objektinstanz festgelegt.
   bei System.Windows.Forms.Control.MarshaledInvoke(Control caller, Delegate method, Object[] args, Boolean synchronous)
   bei System.Windows.Forms.Control.Invoke(Delegate method, Object[] args)
   bei CKAN.ManageMods._UpdateModsList(Dictionary`2 old_modules)
   bei System.Threading.Tasks.Task.Execute()
--- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
   bei System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   bei System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   bei CKAN.ManageMods.<UpdateModsList>d__91.MoveNext()
--- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
   bei System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
```

That function for easy auditing:

https://github.com/KSP-CKAN/CKAN/blob/6ec7270e2fe92845b662ce2ae5d14132a5429590/GUI/Controls/ManageMods.cs#L1033-L1166

**Comment by DasSkelett** [comment]
Hm, `_updateModsList()` didn't change much between the releases.
https://github.com/KSP-CKAN/CKAN/compare/v1.29.2...v1.30.0#diff-b8bc6b0c967d56b2fd678b234d315bf2b61de808100d09bb8c665d3064d6d2ae

Maybe `has_unheld_updates` is evaluted only in the filter invoke, and something in `LabelsHeld()` throws a `NullReferenceException`?

**Comment by MrHuggy** [comment]
Having the exactly the same problem as well. Nothing has change apart for updating CKAN.

**Comment by HebaruSan** [comment]
> Maybe `has_unheld_updates` is evaluted only in the filter invoke, and something in `LabelsHeld()` throws a `NullReferenceException`?

Hmm, `Any` returns `bool`, not an `IEnumerable` that might defer evaluation:

https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.any?view=netframework-4.7.2#System_Linq_Enumerable_Any__1_System_Collections_Generic_IEnumerable___0__System_Func___0_System_Boolean__

**Comment by DasSkelett** [comment]
> Hmm, `Any` returns `bool`, not an `IEnumerable` that might defer evaluation:

It was more about the variable it's called on, `gui_mods' being some enumerable, not what it returns (a bool indeed). And you never know with IEnumerables if it doesn't evaluate even something like `Any()` lazily. But it's a `HashSet` anyway and honestly I don't think that's what happens, was just a stab in the dark. It's the only place in that method where we do something that's more than a simple assignment or call of a framework function and that at least indirectly touches an `Invoke()`.

Or one of the callback `Action`? They also do some `Invokes`, maybe they got lost in the stacktrace.

**Comment by HebaruSan** [comment]
> And you never know with IEnumerables if it doesn't evaluate even something like `Any()` lazily

I believe as a general rule, lazy evaluation can only happen when a function returns either `IEnumerable` or `IEnumerable<>`. Kind of like how `ToList()` prevents lazy evaluation because it returns a `List<>`. So we can be pretty sure about `Any`.

**Comment by Arkerthan** [comment]
Here are my config files. Renamed to txt so I can upload them to github.

[GUIConfig.xml.txt](https://github.com/KSP-CKAN/CKAN/files/6096757/GUIConfig.xml.txt)

[config.json.txt](https://github.com/KSP-CKAN/CKAN/files/6096761/config.json.txt)

**Comment by DasSkelett** [comment]
It is indeed related to the `GUIConfig.xml`. CKAN is hanging when using the GUIConfig from @Arkerthan. The exception stays hidden on Linux, some IDE magic helped:
```
System.NullReferenceException: Object reference not set to an instance of an object
  at CKAN.ManageMods.ApplyHeaderGlyphs () [0x00051] in CKAN/GUI/Controls/ManageMods.cs:1244 
  at CKAN.ManageMods._UpdateFilters () [0x0010b] in CKAN/GUI/Controls/ManageMods.cs:1007 
```
https://github.com/KSP-CKAN/CKAN/blob/76f656b99ef8120511053e3bd099a8ec86a7f29f/GUI/Controls/ManageMods.cs#L1236-L1247

**Comment by HebaruSan** [comment]
So the folks having this problem have all sorted the mod list by the compatibility column?

**Comment by DasSkelett** [comment]
I think so.

@Arkerthan @MrHuggy and anyone else reading this encountering the same bug, a workaround to get CKAN working until we fix the bug:

Edit `<KSP>\CKAN\GUIConfig.xml` and change `KSPCompatibility` to `GameCDompatibility` in the `<SortColumns>` list:
```xml
  <SortColumns>
    <string>KSPCompatibility</string>
  </SortColumns>
```
to
```xml
  <SortColumns>
    <string>GameCompatibility</string>
  </SortColumns>
```
(there might be multiple entries in this list, but you just need to change `KSPCompatibility`)

**Comment by HE4TER** [comment]
The workaround doesn't work for me. SortColumns is empty in my GUIConfig.xml file and changing it to include the GameCompatibility string doesn't get rid of the error or change anything. I also tried deleting the CKAN directory in my KSP installation folder but that also didn't help. Any other way to fix or work around this?

**Comment by HebaruSan** [comment]
@HE4TER, that's **the** work around for this. If it doesn't work for you (and you did it correctly), then you're probably seeing some other poblem.

**Comment by Spotflare** [comment]
Potentially easier workaround (may not work in all instances, but went fine for mine, appears to be occurring again after a fresh install, in case someone else comes looking).

Hit continue on the error, click one of the column headers to sort (I went by name), problem went away and would stop showing the error upon scrolling.

**Comment by HebaruSan** [comment]
@Spotflare, this issue was fixed five years ago in #3319. If you're seeing a similar problem, it should be reported in a new issue.
